### PR TITLE
DEPR: min_periods=None behavior for Rolling.count

### LINKED ISSUE #31302: ENH: Change default behavior of rolling.count to be consistent with others
Following on discussion happened in #30923, we may want to change the default behavior of `rolling.count` with regards to its parameter ``min_periods``, so it is consistent with all other similar APIs such as `rolling.mean` and `rolling.sum`.

#### Code Sample
With the updates from #30923
```python
>>> import numpy as np
>>> import pandas as pd
>>> s = pd.Series([1, 1, 1, np.nan, 1, 1, 1])
>>> s
0    1.0
1    1.0
2    1.0
3    NaN
4    1.0
5    1.0
6    1.0
dtype: float64

# rolling.mean and rolling.sum defaults min_periods to the same value as window size (3 in this case)
# notice that it requires not only the window size to be at least 3, but also the number of valid entries (not NaN) to be at least 3
>>> s.rolling(3).mean()  
0    NaN
1    NaN
2    1.0
3    NaN
4    NaN
5    NaN
6    1.0
dtype: float64

>>> s.rolling(3).sum()
0    NaN
1    NaN
2    3.0
3    NaN
4    NaN
5    NaN
6    3.0
dtype: float64

# the default value of min_periods for rolling.count is 0
# we may want to change this behavior so it's consistent with other APIs
>>> s.rolling(3).count()
0    1.0
1    2.0
2    3.0
3    2.0
4    2.0
5    2.0
6    3.0
dtype: float64

# notice that rolling.count requires window size to be at least equal to min_periods to give a result
# it doesn't care about how many valid entries (not NaN) to determine if it should output NaN
# we should retain this behavior because this function is meant to count the number of valid entries
>>> s.rolling(3, min_periods=3).count()
0    NaN
1    NaN
2    3.0
3    2.0
4    2.0
5    2.0
6    3.0
dtype: float64
```

#### Expected Output
```python
>>> s.rolling(3).count()
0    NaN
1    NaN
2    3.0
3    2.0
4    2.0
5    2.0
6    3.0
dtype: float64
```

#### Problem description
With the updates from #30923, the ``min_periods`` argument of `rolling.count` is now respected (it used to be completely ignored). However, the default value remains 0 for backward compatibility purpose. In future updates we probably want to change this default behavior so it's consistent with other similar APIs.

@mroeschke previously mentioned we needed to start with a ``DeprecationWarning`` to inform users of future changes, then probably in the following release make the actual change.

@jreback @WillAyd 
Let me know what you guys think!

### LINKED ISSUE #35579: BUG: rolling count on FixedForwardWindowIndexer returns NaN instead of 0
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
from pandas.core.window.indexers import FixedForwardWindowIndexer

df = pd.DataFrame([[0, None], [0, None],[0, None],[0, 7]], columns=["a", "b"])
res = df["b"].rolling(window=FixedForwardWindowIndexer(window_size=2),min_periods=0).count()
print(res)
```

#### Problem description

Current output:
0    NaN
1    NaN
2    1.0
3    1.0

#### Expected Output

0    0.0
1    0.0
2    1.0
3    1.0

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.6.13-100.fc30.x86_64
Version          : #1 SMP Fri May 15 00:36:06 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : C.UTF-8
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.0
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 9.0.3
setuptools       : 45.2.0
Cython           : None
pytest           : 4.6.4
hypothesis       : None
sphinx           : 3.1.2
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.2
html5lib         : None
pymysql          : None
psycopg2         : 2.8.5 (dt dec pq3 ext lo64)
jinja2           : 2.10.1
IPython          : 7.15.0
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fsspec           : 0.7.4
fastparquet      : 0.4.1
gcsfs            : 0.6.2
matplotlib       : 3.3.0
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : 0.10.0
pyarrow          : 1.0.0
pytables         : None
pyxlsb           : None
s3fs             : 0.2.2
scipy            : 1.3.1
sqlalchemy       : 1.3.18
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.50.1
</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #31302
- [x] closes #35579 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Additionally refactors `count` to take the `_apply` path to make this op consistent with the others.

### CODE REVIEW

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -471,31 +472,39 @@ def _get_window_indexer(self, window: int) -> BaseIndexer:
...
         """
         if self._selected_obj.ndim == 1:
-            return self._apply_series(homogeneous_func)
+            return self._apply_series(homogeneous_func, name)
 
         obj = self._create_data(self._selected_obj)
+        if name == "count":
+            # GH 12541: Special case for count where we support date-like types
+            obj = notna(obj).astype(int)
+            obj._mgr = obj._mgr.consolidate()
```
Comment (jbrockmendel) [inline pandas/core/window/rolling.py:507]: why is this necessary?

Comment (mroeschke) [reply pandas/core/window/rolling.py:507]: Not 100% sure of best practice regarding consolidation, but I noticed that after this operations the blocks would be unconsolidated. Can remove this is you recommend so.

```
In [7]: df
Out[7]:
           a           b    c
0 2020-01-01  2020-01-01  1.0
1 2020-01-02  2020-01-02  2.0
2 2020-01-03  2020-01-03  3.0

In [8]: notna(df).astype(int)._mgr
Out[8]:
BlockManager
Items: Index(['a', 'b', 'c'], dtype='object')
Axis 1: RangeIndex(start=0, stop=3, step=1)
IntBlock: slice(2, 3, 1), 1 x 3, dtype: int64
IntBlock: slice(0, 1, 1), 1 x 3, dtype: int64
IntBlock: slice(1, 2, 1), 1 x 3, dtype: int64
```

Comment (jbrockmendel) [reply pandas/core/window/rolling.py:507]: It's harmless in this case since this object is never user-facing, just makes an unnecessary copy.  For some ops consolidating improves perf.

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -2050,14 +2034,16 @@ def aggregate(self, func, *args, **kwargs):
...
-            return self._apply(window_func, center=self.center, name="count")
-
+        if self.min_periods is None:
+            warnings.warn(
+                (
+                    "min_periods=None will default to the size of window "
+                    "consistent with other methods in a future version. "
+                    "Specify min_periods=0 instead."
+                ),
+                DeprecationWarning,
```
Comment (jreback) [inline pandas/core/window/rolling.py:2044]: make this a FutureWarning. I know its loud but i think we need to.

Comment (jreback) [reply pandas/core/window/rolling.py:2044]: alt we could do this for a version and then change it.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by phofl** [comment]
This will close #35579 too.

You could use
```
def test_fixed_forward_indexer_count():
    # GH: 35579
    df = DataFrame({"b": [None, None, None, 7]})
    indexer = FixedForwardWindowIndexer(window_size=2)
    result = df.rolling(window=indexer, min_periods=0).count()
    expected = DataFrame({"b": [0.0, 0.0, 1.0, 1.0]})
    tm.assert_frame_equal(result, expected)
```
as a test.

**Comment by mroeschke** [comment]
Thanks @phofl. Added that unit test to this PR

File: pandas/core/window/rolling.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/window/rolling.py]: Changed to FutureWarning

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke
